### PR TITLE
New Metadata API Defaults & IIIF Collection Updates for 1.4.0

### LINKED ISSUE #267: Updates to the IIIF Collection Manifest Template
# What's needed?
- Updates to the IIIF Collection Manifest Template to change to use the metadata exposed endpoint Route to generate URL instead of entity node canonical, so aliases don't break the links for children objects

## Changes need to be applied:

- [x] to corresponding metadata display entity config file
- [x] to JSON-encoded version of the template

Per @DiegoPino:

```
 {#
Updated 06/13/2024 for 1.4.0

Uses metadata exposed endpoint Route to generate URL so Aliases don't break it

This Template generates IIIF 3.0 API Manifest for Collections using a Drupal Views
JSON to be used by a Drupal Views (REST) to generate a simpler to parse/join listing of Canvases
For a Collection Manifest Manifest

Structures passed from Archipelago are
- node
  The full Node entity
- data
  The Full JSON Structure converted in an Array
- iiif_server
  The publicly available IIIF Server (Global Setting)

Since JSON was decoded into the 'data' values are valid PHP but could not
be valid JSON when outputting. So We use data.description|json_encode|raw
Without Double quotes.
#}
{% set Webserverurl =  url('<front>')|render|replace({':8001/':''}) ~ ':8183/iiif/2/' %}
{% set IIIFserverurl = iiif_server ? iiif_server ~ '/': Webserverurl %}
{% set currenturl = url('<current>')|render %}
{% set nodeid = 1 %} 
{% if node.id %}
{# this is needed to the View Argument can conform to the format even when no node exists when saving the template #}
  {% set nodeid = node.id %}
  {% set nodeurl = url('entity.node.canonical', {'node': node.id}, {'absolute': true}) %}
{% else %}
  {% set nodeid = 1 %} 
  {% set nodeurl = currenturl %}
{% endif%}
{% set len = language.getId() ? language.getId() : "en" %}
{% set computed_rights = "http://rightsstatements.org/vocab/InC-EDU/1.0/" %}
{% if data.rights|length > 0 or data.rights_statements|length > 0 %}
	{% if data.rights|length > 0 %} 
		{% set computed_rights = data.rights %}
	{% endif %}
{% elseif data.rights_statements|length > 0 %}
	{% if data.rights_statements in ['In Copyright'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/InC/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['In Copyright - EU Orphan Work'] %} 
    	{% set computed_rights = "http://rightsstatements.org/vocab/InC-OW-EU/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['In Copyright - Educational Use Permitted'] %} 
	   	{% set computed_rights = "http://rightsstatements.org/vocab/InC-EDU/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['In Copyright - Non-Commercial Use Permitted'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/InC-NC/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['In Copyright - Rights-holder(s) Unlocatable or Unidentifiable'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/InC-RUU/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['No Copyright - Contractual Restrictions'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/NoC-CR/1.0/" %}
	{% endif %} 
	{% if data.rights_statements in ['No Copyright - Non-Commercial Use Only'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/NoC-NC/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['No Copyright - Other Known Legal Restrictions'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/NoC-OKLR/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['No Copyright - United States'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/NoC-US/1.0/" %}
	{% endif %} 
	{% if data.rights_statements in ['Copyright Not Evaluated'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/CNE/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['Copyright Undetermined'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/UND/1.0/" %}
	{% endif %}
	{% if data.rights_statements in ['No Known Copyright'] %} 
		{% set computed_rights = "http://rightsstatements.org/vocab/NKC/1.0/" %}
	{% endif %}
{% endif %}
{
{# Top Object Metadata #}
"@context": [
  "http://www.w3.org/ns/anno.jsonld",
  "http://iiif.io/api/presentation/3/context.json"
],
"type": "Collection",
"id": "{{ currenturl }}",
"label": {
"{{ len }}": [
{{ node.label|json_encode|raw }}
]
},
"requiredStatement": {
"label": { "en": [ "Attribution" ] },
"value": { "en": [ "Provided by Archipelago Deployment" ] }
},
"rights": {{ computed_rights|json_encode|raw }},
"metadata": [
{% for agent in attribute(data, 'edm_agent') %}
    {
    "label": {
    "{{ len }}": [
    {{ agent.role_label|json_encode|raw }}
    ]
    },
    "value": {
    "{{ len }}": [
    {{ agent.name_label|json_encode|raw }}
    ]
    }
    },
{% endfor %}
{
"label": {
"{{ len }}": [
"description"
]
},
"value": {
"{{ len }}": [
{{ data.description|json_encode|raw }}
]
}
},
{
"label": {
"{{ len }}": [
"navDate"
]
},
"value":  {
"{{ len }}": [
"{{ node.getChangedTime|date("F jS \\a\\t g:ia") }}"
]
}
},
{
"label": {
"{{ len }}": [
"license"
]
},
"value": {
"{{ len }}": [
{{ computed_rights|json_encode|raw }}
]
}
}
],
{% if attribute(data, 'memberof') %}
    {% set currentnode = 'node/' ~ data.memberof %}
    "within": "{{ url(currentnode) }}",
{% else %}
    {% set currentnode = '<current>' %}
{% endif %}
{# fetch all children using a View and attach them as Items #}
  "items": [
{% set collection_view = drupal_view('data_collection_manifest','children_ismemberof_manifest_rest', nodeid)|render %}
{% set collection_items = collection_view|sbf_json_decode %}
{% for items in collection_items %}
{# canvas|json_encode #}
    {
    {% if 'CreativeWorkSeries' in items['ado_type']|split(',') %}
        "id": "{{ url('format_strawberryfield.metadatadisplay_caster', {'node': items['uuid'],'metadataexposeconfig_entity':'iiifmanifest3cws', 
        'format':'default.jsonld'}, {'absolute': true, 'alias': true}) }}",
        "type": "Manifest",
     {% elseif 'Collection' in items['ado_type']|split(',') %}
        "id": "{{ url('format_strawberryfield.metadatadisplay_caster', {'node': items['uuid'], 'metadataexposeconfig_entity':'iiifmanifest3collection', 
        'format':'default.jsonld'}, {'absolute': true, 'alias': true}) }}",
        "type": "Collection",
    {% else %}
         "id": "{{ url('format_strawberryfield.metadatadisplay_caster', {'node': items['uuid'], 'metadataexposeconfig_entity':'iiifmanifest', 
        'format':'default.jsonld'}, {'absolute': true, 'alias': true}) }}",
        "type": "Manifest",
    {% endif %}
       "label": { "{{len}}": [ "{{ items['label'] }}" ] },
       {{ items['thumbnail']|raw }}
    }
    {{ not loop.last ? ',' : '' }}
{% endfor %}
  ]
}
```

### LINKED ISSUE #268: Provide example Config/Views for new Metadata API endpoint
# What?

@alliomeria probably a DC based` OAI-PMH` one. Means also templates/Views/config. Documentation should be at least be a boiler plate with the basics?

### LINKED ISSUE #270: 1.4.0 Macro issue: Release tasks pending
# What?

This is a macro-issue, self reminder of the release tasks pending and needed before going public. 

- [ ] Test and Merge Open Pull requests against 1.4.0 and 0.8.0 for all Modules. 
- [ ] Deploy an 1.3.0, apply Manually 1.4.0/0.8.0 Branches and Update everything to latest. Document so we can have an Upgrade path document for people coming from 1.3.0.
- [ ] Updated Twig templates and add {# Info #} about the changes. Includes twig.html and also export/ready to import entities (IIIF Collection Manifest & Open API Item & Wrapper for DC templates done via https://github.com/esmero/archipelago-deployment/pull/271) 
- [ ] Update Archipelago Base theme. The `Colors` module is something we could get rid of during this release and use Barrio Colors instead BUT it might affect users that are already overriding the Base theme via Asset Injector. TBD
- [ ] Update/Save Theme settings, make sure that the Colors Cache (if we keep it) is also updated (not git tracked)
- [ ] Import all new Configs (right now all over the place, studio.esmero.io, local deployments) 
  - [ ] Import all new Solr Index, Fields, Field Types (ML), Facets and Facet Summaries
  - [ ] Import all new Views, Update existing ones. Advanced Search and Data/driving Autocompletes
  - [ ] Import and setup all Blocks. Including Exposed ones from Forms and Facets
  - [ ] Import Configs/Updates for Strawberry Runners
  - [ ] Import Configs/Updates for General ones (IIIF general Settings, Archipelago File based ones
  - [x] Generate an example for OpenAPI (means 2 templates + View + Config)
  - [ ] Update AMI set (demo one) with an example with Audio/Video and two Subtitles
  - [ ] Updated all Mirador Settings/View Modes to use 4.0.0
  - [ ] Test/double test Webforms and their View (autocomplete) integrations
- [ ] Export composer.json & composer.lock & config/sync folder and make a pull to this module and `archipelago-deployment-live`
- [ ] Merge and build a new multi platform NLP Container. Export to docker hub
- [ ] Merge and build a new multi platform PHP Container (using latest supported by our code). Export to docker hub
- [ ] Merge and build a new multi platform PHP Dev Container (using latest supported by our code). Export to docker hub
- [ ] Update docker-compose files everywhere.
- [x] https://github.com/esmero/archipelago-deployment/issues/272
- [ ] Using all of this Test (Diego and Allison) locally (do we need a list of things to test here?)

@alliomeria I will update this list/add more stuff during what is left of this week. Danke

### PR DESCRIPTION
**Pull request by alliomeria**
For #268:
- Metadata API default View
- Metadata APi default Configuration .yml
- OAI-PMH Item template
- OAI-PMH Wrapper template for DC
- updated deploy.sh script (thanks for the reminder Diego!)

For #267 
- updated versions of the IIIF Collection Template

### CODE REVIEW

**Comment by DiegoPino** [comment]
@alliomeria this is great.
Could you please also add the new two Twig template JSONs to https://github.com/esmero/archipelago-deployment/blob/1.3.0/scripts/archipelago/deploy.sh [and updated one?](https://github.com/esmero/archipelago-deployment/blob/1.3.0/scripts/archipelago/update_deployed.sh). That way these get ingested by default on install

Really good work!

**Comment by alliomeria** [comment]
Yes, will do that right now. Thanks for your feedback @DiegoPino !

**Comment by alliomeria** [comment]
Done @DiegoPino --let me know if anything else is needed for this.

**Comment by DiegoPino** [comment]
@alliomeria the the update_deployed.sh too please!

File: d8content/metadatadisplay_entity_15.json [hunk 1]
```diff
@@ -1,12 +1 @@
...
...r\n   -%}\r\n {%- else -%}\r\n {# No results so no resumption token #}\r\n  {%- set resumptiontoken = '' -%}\r\n {%- endif -%}\r\n <request {%- for var,value in data_api.query %} {{var}}=\"{{value}}\" {% endfor -%}>{{ resumptiontoken }}<\/request>\r\n {%- if data_api.query.verb == \"ListSets\" -%} \r\n\t\t<error code=\"noSetHierarchy\">This repository does not support sets<\/error>\r\n {%- else -%}\r\n   {%- if data_api.query.verb -%}\r\n <{{ data_api.query.verb }}>\r\n   {%- else -%}\r\n <GetRecord>\r\n   {%- endif -%} \r\n   {%- for XML_doc in data -%}\r\n     {%- if data_api.query.verb == \"ListRecords\" -%}\r\n <record>\r\n     {%- endif -%}\r\n   {{ XML_doc }}\r\n     {%- if data_api.query.verb == \"ListRecords\" -%}  \r\n <\/record>\r\n     {%- endif -%}\r\n   {% endfor %}\r\n {#   {{ data_api|json_encode|raw }} #}\r\n  {#  {{ data_api_context|json_encode|raw }} #}\r\n   {%- if data_api.query.verb -%}\r\n <\/{{ data_api.query.verb }}>\r\n   {%- else -%}\r\n <\/GetRecord>\r\n   {%- endif -%}\r\n {%- endif -%}\r\n<\/OAI-PMH>","name":"OAI-PMH Wrapper","langcode":"und","revision_log_message":null,"drupal_internal__vid":17},"type":"metadatadisplay_entity--metadatadisplay_entity"}}
```
Comment (DiegoPino) [inline d8content/metadatadisplay_entity_15.json]: @alliomeria heys. This needs to keep the old UUID. Which was 9be8473c-3fb3-42ee-aef2-431b2d4f93bb.

File: d8content/metadatadisplay_entity_18.json [hunk 2]
```diff
@@ -0,0 +1 @@
...r\n{% if node.id %}\r\n    {% set nodeurl = url('entity.node.canonical', {'node': node.id}, {'absolute': true}) %}\r\n{% else %}\r\n    {% set nodeurl = '<current>' ~ '\/iiif\/manifest' %}\r\n{% endif%}\r\n {% if attribute(data, 'ismemberof') %}\r\n {% if attribute(data, 'ismemberof') is iterable %}\r\n {% for collection in attribute(data, 'ismemberof')%}\r\n   {% set parentnode = bamboo_load_entity('node', collection) %}\r\n   {% set parentnodeurl = url('entity.node.canonical', {'node': collection}, {'absolute': true}) %}\r\n<dc:relation>{{ parentnode.title.value }}<\/dc:relation>\r\n{% endfor %}\r\n{% else %}\r\n  {% set parentnode = bamboo_load_entity('node', attribute(data, 'ismemberof')) %}\r\n  {% set parentnodeurl = url('entity.node.canonical', {'node': attribute(data, 'ismemberof')}, {'absolute': true}) %}\r\n  <dc:relation>{{ parentnode.title.value }}<\/dc:relation>\r\n  {% endif %}\r\n{% endif %}\r\n{# Rights #}\r\n{% if data.rights|length > 0 %} \r\n  <dc:rights>{{ data.rights }}<\/dc:rights>\r\n{% endif %}\r\n<\/oai_dc:dc>\r\n{% endif %}\r\n<\/metadata>\r\n{% endif %}"},"id":"b5c12175-e0a9-49e9-a698-cbba342667e0","type":"metadatadisplay_entity--metadatadisplay_entity"}}
```
Comment (DiegoPino) [inline d8content/metadatadisplay_entity_18.json]: @alliomeria I think we might have trouble here with the "revision" data ... pretty sure JSON API will complain.
Could you manually remove 
"revision_created":"2024-03-08T16:55:15+00:00" and "drupal_internal__vid":18 ? Same applies to the other one.
The scripts (export import one) was created before I built revisions and I will have to update it for this release too. Sorry

File: scripts/archipelago/update_deployed.sh [hunk 3]
```diff
@@ -65,6 +65,14 @@ echo -e "IIIF 3.0 Thumbnail Manifest Snippet for Collections\n"
 
 curl -w "\n" --user jsonapi:jsonapi -H 'Accept: application/vnd.api+json' -H 'Content-type: application/vnd.api+json' -XPATCH http://esmero-web/jsonapi/metadatadisplay_entity/metadatadisplay_entity/adf12524-11ca-4925-bc23-d028556e75ac --data-binary @d8content/metadatadisplay_entity_16.json;
 
+echo -e "OAI-PMH Wrapper\n"
```
Comment (DiegoPino) [inline scripts/archipelago/update_deployed.sh:68]: Cool!

**Review by DiegoPino — CHANGES_REQUESTED** [review]
Small changes sorry. See my comments, please ask for clarification

File: d8content/metadatadisplay_entity_18.json (hunk 2 above)
Comment (alliomeria) [reply d8content/metadatadisplay_entity_18.json]: Ok, think all the extra new Revision & drupal_internal_vid keys are removed for these now. Understood about that script needing updates...

**Review by DiegoPino — APPROVED** [review]
Wonderful @alliomeria . Thanks for addressing my code reviews and fixing all ASAP. Ready for merging now!

**Comment by DiegoPino** [comment]
Done! Woohoo!

**Pull request merged**